diff -Nur linux-2.6.27.official/drivers/mtd/maps/Kconfig linux-2.6.27/drivers/mtd/maps/Kconfig
--- linux-2.6.27.official/drivers/mtd/maps/Kconfig	2008-10-10 06:13:53.000000000 +0800
+++ linux-2.6.27/drivers/mtd/maps/Kconfig	2009-04-29 22:11:33.000000000 +0800
@@ -128,6 +128,14 @@
 	  demonstration board. If you have one of these boards and would like
 	  to use the flash chips on it, say 'Y'.
 
+config MTD_CAMELOT
+	tristate "Flash device mapped on Camelot"
+	select MTD_COMPLEX_MAPPINGS 
+	select MTD_ROM
+	select MTD_PARTITIONS
+	help
+	  This enables access routines for the flash chips on the Camelot board.
+
 config MTD_TS5500
 	tristate "JEDEC Flash device mapped on Technologic Systems TS-5500"
 	depends on X86
diff -Nur linux-2.6.27.official/drivers/mtd/maps/Makefile linux-2.6.27/drivers/mtd/maps/Makefile
--- linux-2.6.27.official/drivers/mtd/maps/Makefile	2008-10-10 06:13:53.000000000 +0800
+++ linux-2.6.27/drivers/mtd/maps/Makefile	2009-04-29 22:11:33.000000000 +0800
@@ -65,3 +65,4 @@
 obj-$(CONFIG_MTD_OMAP_NOR)	+= omap_nor.o
 obj-$(CONFIG_MTD_MTX1)		+= mtx-1_flash.o
 obj-$(CONFIG_MTD_INTEL_VR_NOR)	+= intel_vr_nor.o
+obj-$(CONFIG_MTD_CAMELOT)	+= camelot_flash.o
diff -Nur linux-2.6.27.official/drivers/watchdog/Kconfig linux-2.6.27/drivers/watchdog/Kconfig
--- linux-2.6.27.official/drivers/watchdog/Kconfig	2008-10-10 06:13:53.000000000 +0800
+++ linux-2.6.27/drivers/watchdog/Kconfig	2009-04-29 22:11:35.000000000 +0800
@@ -704,6 +704,12 @@
 	help
 	  Hardware driver for the built-in watchdog timer on TXx9 MIPS SoCs.
 
+config CAMELOT_WDT
+	tristate "Camelot Watchdog Timer"
+	depends on MIPS_CAMELOT
+	help
+	  Hardware driver for the built-in watchdog timer on Camelot SoCs.
+
 # PARISC Architecture
 
 # POWERPC Architecture
diff -Nur linux-2.6.27.official/drivers/watchdog/Makefile linux-2.6.27/drivers/watchdog/Makefile
--- linux-2.6.27.official/drivers/watchdog/Makefile	2008-10-10 06:13:53.000000000 +0800
+++ linux-2.6.27/drivers/watchdog/Makefile	2009-04-29 22:11:35.000000000 +0800
@@ -102,6 +102,7 @@
 obj-$(CONFIG_SIBYTE_WDOG) += sb_wdog.o
 obj-$(CONFIG_AR7_WDT) += ar7_wdt.o
 obj-$(CONFIG_TXX9_WDT) += txx9wdt.o
+obj-$(CONFIG_CAMELOT_WDT) += camelot_wdt.o
 
 # PARISC Architecture
 
