--- uboot/net/net.c	2010-04-21 10:07:55.513869740 +0800
+++ uboot/net/net.c	2010-04-21 10:23:24.820117346 +0800
@@ -202,7 +202,9 @@
 =======================================*/
 #define RL_DEBUG 0
 int EMERGENCY_MODE=0;
-uchar	*NetTcpDestMAC;	/* MAC address of destination	*/
+/* EZP: do not use unallocated pointer, use array instead. */
+/*uchar	*NetTcpDestMAC; */	/* MAC address of destination	*/
+uchar    NetTcpDestMAC[12];
 
 static unsigned int seq_num __attribute__ ((packed));
 static unsigned int ack_num __attribute__ ((packed));
@@ -217,8 +219,8 @@
 }  TCPfakehdr;
 
 TCPfakehdr fh;
-
-int Listen_FG;
+/* EZP: set Listen_FG initial value */
+int Listen_FG = 0;
 int sendresult_flag;   //0:start 1:fail 2:reset
 
 uchar *httpdReplyData; //httpd reply web page to client
@@ -1848,7 +1850,8 @@
 		return;
 	}
 	//Check Run image length
-	if ( ntohl(fwHeader->ih_size) != (HTTP_FileLen-sizeof(image_header_t)))
+	/* EZP: image size may less than or equal to transfered data size. */
+	if ( ntohl(fwHeader->ih_size) > (HTTP_FileLen-sizeof(image_header_t)))
 	{
 		puts("Check Image Length error\n\n");
 		sendresult_flag=1;//fail
@@ -1908,8 +1911,8 @@
 			 	ack_num =(ntohl(tcpPkt->tcp_seq)+ tcp_data_len);
 			}
 			#endif
-			
-			NetTcpDestMAC = NULL;
+		    /* EZP: we used array instead */
+			/* NetTcpDestMAC = NULL; */
 			break;
 		#endif
 		} 
--- uboot/include/net.h	2010-04-21 10:07:55.493869645 +0800
+++ uboot/include/net.h	2010-04-21 10:20:51.264359083 +0800
@@ -445,7 +445,10 @@
 
 //RL added HTTPD define
 extern int EMERGENCY_MODE; //check if into EMERGENCY MODE
-extern uchar *NetTcpDestMAC; // MAC address of destination
+/* EZP: do not use unallocated pointer, use array instead. */
+/* extern uchar *NetTcpDestMAC; */// MAC address of destination
+extern uchar    NetTcpDestMAC[];
+
 extern uchar *httpdReplyData; //httpd reply web page to client
 extern int Listen_FG;
 extern int sendresult_flag;   //0:start 1:fail 2:reset
--- uboot/net/httpd.c	2010-04-19 19:08:23.742618860 +0800
+++ uboot/net/httpd.c	2010-04-19 19:08:12.932618689 +0800
@@ -30,7 +30,7 @@
 *********************************************/
 int            HTTP_FileLen;
 int            MaxLen_flag;
-#define        MAX_FILESIZE  (4*1024*1024 )
+#define        MAX_FILESIZE  (8*1024*1024 )
 
 unsigned int strtoint(char *str, unsigned int count);
 //char 		*file;
