--- beceemwimax/tgtxvi020/Defines.h	2011-01-31 18:54:03.000000000 +0800
+++ beceemwimax/tgtxvi020/Defines.h	2011-02-08 14:25:23.772255797 +0800
@@ -53,7 +53,7 @@
 #define INTERFACE_TYPE_ETHERNET		0x0001
 #define INTERFACE_TYPE_SERIAL		0x0002
 #define WAIT_TIME_OUT					(1000000)
-#define INTERFACE_CONFIG_FILE "RemoteProxy.cfg"
+#define INTERFACE_CONFIG_FILE "/etc/RemoteProxy.cfg"
 
 
 #define PACKET_SIGNATURE			(0xBABEFACE)
