--- lighttpd-1.4.30/src/mod_webdav.c	2013-07-16 18:28:44.649071988 +0800
+++ lighttpd-1.4.30/src/mod_webdav.c	2013-07-16 18:28:32.450270755 +0800
@@ -1780,6 +1780,7 @@
 			}
 		} else {
 			con->http_status = 201;
+                        sync();
 			con->file_finished = 1;
 		}
 
@@ -1971,6 +1972,7 @@
 						return HANDLER_FINISHED;
 					} else {
 						con->http_status = 201; /* created */
+                                                sync();
 					}
 				} else {
 					con->http_status = 200; /* modified */
@@ -2163,6 +2165,7 @@
 				rmdir(con->physical.path->ptr);
 			}
 			con->http_status = 201;
+                        sync();
 			con->file_finished = 1;
 		} else {
 			/* it is just a file, good */
@@ -2189,6 +2192,7 @@
 
 			if (-1 == r) {
 				con->http_status = 201; /* we will create a new one */
+                                sync();
 				con->file_finished = 1;
 
 				switch(errno) {
