diff -Naur linux-2.6.24.4/Makefile linux-2.6.24.4-oxnas/Makefile
--- linux-2.6.24.4/Makefile	2008-03-25 02:49:18.000000000 +0800
+++ linux-2.6.24.4-oxnas/Makefile	2009-05-08 16:27:10.000000000 +0800
@@ -190,9 +190,8 @@
 # Default value for CROSS_COMPILE is not to prefix executables
 # Note: Some architectures assign CROSS_COMPILE in their arch/*/Makefile
 export KBUILD_BUILDHOST := $(SUBARCH)
-ARCH		?= $(SUBARCH)
-CROSS_COMPILE	?=
-
+ ARCH		?= arm
+ CROSS_COMPILE	?= arm-linux-uclibcgnueabi-
 # Architecture as present in compile.h
 UTS_MACHINE 	:= $(ARCH)
 SRCARCH 	:= $(ARCH)
