diff -Naur snort-2.9.2.2.orig/configure snort-2.9.2.2/configure
--- snort-2.9.2.2.orig/configure	2012-03-29 11:43:47.000000000 -0500
+++ snort-2.9.2.2/configure	2012-03-29 11:46:57.000000000 -0500
@@ -13043,7 +13043,7 @@
 $as_echo "#define LINUX 1" >>confdefs.h
 
 
-    extra_incl="-I/usr/include/pcap"
+    extra_incl=""
     ;;
   *-hpux10*|*-hpux11*)
 
@@ -14549,40 +14549,7 @@
   fi
 fi
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for pcap_lex_destroy" >&5
-$as_echo_n "checking for pcap_lex_destroy... " >&6; }
-if test "$cross_compiling" = yes; then :
-  { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
-See \`config.log' for more details" "$LINENO" 5; }
-else
-  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-#include <pcap.h>
-
-int
-main ()
-{
-
-   pcap_lex_destroy();
-
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_run "$LINENO"; then :
-  have_pcap_lex_destroy="yes"
-else
   have_pcap_lex_destroy="no"
-fi
-rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
-  conftest.$ac_objext conftest.beam conftest.$ac_ext
-fi
-
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $have_pcap_lex_destroy" >&5
-$as_echo "$have_pcap_lex_destroy" >&6; }
 if test "x$have_pcap_lex_destroy" = "xyes"; then
 
 $as_echo "#define HAVE_PCAP_LEX_DESTROY 1" >>confdefs.h

