--- old/drivers/char/ralink_gpio.h	2013-12-20 10:18:04.589336928 +0800
+++ new/drivers/char/ralink_gpio.h	2013-12-20 10:16:42.373338201 +0800
@@ -36,6 +36,7 @@
 
 #include <asm/rt2880/rt_mmap.h>
 
+#include <linux/ezp-gpio.h>
 
 #if defined (CONFIG_RALINK_RT3052)
 #define RALINK_GPIO_HAS_5124            1
@@ -538,3 +539,33 @@
 #define RALINK_GPIO(x)			(1 << x)
 
 #endif
+
+#define RALINK_GPIO_0           0x00000001
+#define RALINK_GPIO_1           0x00000002
+#define RALINK_GPIO_2           0x00000004
+#define RALINK_GPIO_3           0x00000008
+#define RALINK_GPIO_4           0x00000010
+#define RALINK_GPIO_5           0x00000020
+#define RALINK_GPIO_6           0x00000040
+#define RALINK_GPIO_7           0x00000080
+#define RALINK_GPIO_8           0x00000100
+#define RALINK_GPIO_9           0x00000200
+#define RALINK_GPIO_10          0x00000400
+#define RALINK_GPIO_11          0x00000800
+#define RALINK_GPIO_12          0x00001000
+#define RALINK_GPIO_13          0x00002000
+#define RALINK_GPIO_14          0x00004000
+#define RALINK_GPIO_15          0x00008000
+#define RALINK_GPIO_16          0x00010000
+#define RALINK_GPIO_17          0x00020000
+#define RALINK_GPIO_18          0x00040000
+#define RALINK_GPIO_19          0x00080000
+#define RALINK_GPIO_20          0x00100000
+#define RALINK_GPIO_21          0x00200000
+#define RALINK_GPIO_22          0x00400000
+#define RALINK_GPIO_23          0x00800000
+#define RALINK_GPIO(x)          (1 << x)
+
+#define BUTTON_SD_CARD      21
+#define BUTTON_FUNC         0x60
+
--- old/drivers/char/ralink_gpio.c	2013-12-20 10:18:04.589336928 +0800
+++ new/drivers/char/ralink_gpio.c	2013-12-20 10:12:20.437342259 +0800
@@ -44,7 +44,6 @@
 #endif
 #include <asm/uaccess.h>
 #include "ralink_gpio.h"
-#include <linux/ezp-gpio.h>
 #include <asm/rt2880/surfboardint.h>
 
 #ifdef  CONFIG_DEVFS_FS
@@ -52,6 +51,16 @@
 static  devfs_handle_t devfs_handle;
 #endif
 
+/* EZP: set button GPIO definitions */
+#ifdef CONFIG_RALINK_MT7620
+    #warning "MT7620"
+    #define RALINK_GPIO_BUTTON_SD_CARD RALINK_GPIO_21
+#endif
+
+/* EZP: set button GPIO definitions */
+int ezp_gpio_start=0;
+int sd_card = 1, sd_count = 0, sd_done = 1;
+
 #define NAME			"ralink_gpio"
 #define RALINK_GPIO_DEVNAME	"gpio"
 int ralink_gpio_major = 252;
@@ -118,7 +127,6 @@
 u32 ra_gpio3924_led_clr = 0;
 u32 ra_gpio7140_led_set = 0;
 u32 ra_gpio7140_led_clr = 0;
-int ezp_gpio_start=0;
 #if defined (RALINK_GPIO_HAS_7224)
 u32 ra_gpio72_led_set = 0;
 u32 ra_gpio72_led_clr = 0;
@@ -132,6 +140,7 @@
 u32 ra_gpio9564_led_set = 0;
 u32 ra_gpio9564_led_clr = 0;
 #endif
+
 struct ralink_gpio_led_status_t {
 	int ticks;
 	unsigned int ons;
@@ -141,6 +150,11 @@
 } ralink_gpio_led_stat[RALINK_GPIO_NUMBER];
 #endif
 
+struct work_button {
+    int  button_helper_data;
+    struct work_struct button_helper;
+}button_data;
+extern int call_usermodehelper(char *path, char **argv, char **envp, int wait);
 MODULE_DESCRIPTION("Ralink SoC GPIO Driver");
 MODULE_AUTHOR("Winfred Lu <winfred_lu@ralinktech.com.tw>");
 MODULE_LICENSE("GPL");
@@ -910,6 +924,29 @@
 	open:		ralink_gpio_open,
 	release:	ralink_gpio_release,
 };
+void button_helper_action(struct work_struct *work)
+{
+    int retval;
+    char *gpio_argv[3];
+    static char *gpio_envp[] =
+    { "HOME=/", "PATH=/sbin:/usr/sbin:/bin:/usr/bin", NULL };
+    struct work_button *button_data =
+        container_of(work, struct work_button, button_helper);
+    char function_id[6];
+
+    gpio_argv[0] = "/sbin/ezp-gpio";
+    if (button_data->button_helper_data == BUTTON_SD_CARD) {
+        gpio_argv[1] = "sdcard";
+        gpio_argv[2] = NULL;
+    } else {
+        printk("no action!\n");
+        return;
+    }
+    retval = call_usermodehelper(gpio_argv[0], gpio_argv, gpio_envp, 0);
+    if (retval) {
+        printk("call_usermodehelper retval %d\n", retval);
+    }
+}
 
 #ifdef CONFIG_RALINK_GPIO_LED
 
@@ -959,6 +996,7 @@
 {
 	int i;
 	unsigned int x;
+    unsigned long tmp;
 
 #if defined (RALINK_GPIO_HAS_2722)
 	for (i = 0; i < 22; i++) {
@@ -2135,6 +2173,32 @@
 	ra_gpio9564_led_set = ra_gpio9564_led_clr = 0;
 #endif
 
+    /* EZP: read gpio status */
+    tmp = le32_to_cpu(*(volatile u32 *)(RALINK_REG_PIODATA));
+    if (tmp & RALINK_GPIO_BUTTON_SD_CARD) {
+        if (!sd_card) {
+            sd_count=0;
+            sd_done=0;
+        }
+        sd_card=1;
+    } else {
+        if (sd_card) {
+            sd_count=0;
+            sd_done=0;
+        }
+        sd_card=0;
+    }
+    if (!sd_done) sd_count++;
+    if (sd_count > 10) {
+        sd_done=1;
+        sd_count=0;
+        button_data.button_helper_data = BUTTON_SD_CARD;
+        schedule_work(&button_data.button_helper);
+    }
+#ifdef RALINK_GPIO_BUTTON_FUNC
+    ezp_gpio_handle_func_button(tmp,RALINK_GPIO_BUTTON_FUNC_MASK);
+#endif /* RALINK_GPIO_BUTTON_FUNC */
+
 	init_timer(&ralink_gpio_led_timer);
 	ralink_gpio_led_timer.expires = jiffies + RALINK_GPIO_LED_FREQ;
 	add_timer(&ralink_gpio_led_timer);
@@ -2248,6 +2312,9 @@
 #ifdef CONFIG_RALINK_GPIO_LED
 	ralink_gpio_led_init_timer();
 #endif
+    /* Initialize the button helper. */
+    INIT_WORK(&button_data.button_helper, button_helper_action);
+       
 	printk("Ralink gpio driver initialized\n");
 	return 0;
 }
